flutter add-to-app module aar build failure exit code not propagated

we are building a flutter add-to-app module on our CI server but when the flutter build fails, it does not fail the aar build. So the CI server says the build has succeeded while it actually failed.

## Steps to Reproduce

1. create add-to-app module
2. make sure the build fails (introduce compile error in dart code)
3. build the android artifact for add-to-app module

**Expected results:**
I expected to build to fail , i.e. return a non-zero exit code

**Actual results:** 
The output tells me the build has failed but the exit code is 0 (see logs)

<details>
  <summary>Logs</summary>

```
C:\system9\flutter\kwsc_flutter_module>flutter build aar  --no-profile --no-release
Running Gradle task 'assembleAarDebug'...
Running Gradle task 'assembleAarDebug'... Done                     12,0s

lib/injection_container.dart:104:9: Error: No named parameter with the name 'registrationFingerprint'.
        registrationFingerprint: credentials['secret'],
        ^^^^^^^^^^^^^^^^^^^^^^^
../auth_classic/lib/domain/entities/classic_credentials.dart:19:3: Context: Found this candidate, but the arguments don't match.
  ClassicCredentials({
  ^^^^^^^^^^^^^^^^^^


FAILURE: Build failed with an exception.

* Where:
Script 'C:\programs\flutter\packages\flutter_tools\gradle\flutter.gradle' line: 904

* What went wrong:
Execution failed for task ':flutter:compileFlutterBuildDebug'.
> Process 'command 'C:\programs\flutter\bin\flutter.bat'' finished with non-zero exit value 1

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 11s


Gradle task assembleAarDebug failed with exit code 0. 
```

```
Flutter 1.22.4 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 1aafb3a8b9 (3 weeks ago) • 2020-11-13 09:59:28 -0800
Engine • revision 2c956a31c0
Tools • Dart 2.10.4

C:\system9\flutter\kwsc_flutter_module>flutter doctor -v
[√] Flutter (Channel stable, 1.22.4, on Microsoft Windows [Version 10.0.18363.1198], locale nl-BE)
    • Flutter version 1.22.4 at C:\programs\flutter
    • Framework revision 1aafb3a8b9 (3 weeks ago), 2020-11-13 09:59:28 -0800
    • Engine revision 2c956a31c0
    • Dart version 2.10.4
    • Pub download mirror http://droid:4000


[√] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at C:\programs\Android\sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_SDK_ROOT = C:\programs\Android\sdk
    • Java binary at: C:\programs\Java\jdk8u162_64bit\bin\java
    • Java version Java(TM) SE Runtime Environment (build 1.8.0_162-b12)
    • All Android licenses accepted.

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/windows#android-setup for detailed instructions).

[!] Connected device
    ! No devices available

! Doctor found issues in 2 categories.
```

</details>

I can't seem to reproduce

```
> flutter build aar
Running Gradle task 'assembleAarDebug'...
Running Gradle task 'assembleAarDebug'... Done                     24.8s

lib/main.dart:16:1: Error: 'make' isn't a type.
make it crash
^^^^
lib/main.dart:16:6: Error: Expected ';' after this.
make it crash
     ^^
lib/main.dart:16:9: Error: 'crash' isn't a type.
make it crash
        ^^^^^
lib/main.dart:17:3: Error: Expected ';' after this.
  runApp(
  ^^^^^^
lib/main.dart:21:6: Error: Expected ')' before this.
    ),
     ^

Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: /Users/xster/.pub-cache/hosted/pub.dartlang.org/espresso-0.0.1+7/android/src/main/java/androidx/test/espresso/flutter/EspressoFlutter.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

FAILURE: Build failed with an exception.

* Where:
Script '/Users/xster/development/flutter/packages/flutter_tools/gradle/flutter.gradle' line: 900

* What went wrong:
Execution failed for task ':flutter:compileFlutterBuildDebug'.
> Process 'command '/Users/xster/development/flutter/bin/flutter'' finished with non-zero exit value 1

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 23s


Gradle task assembleAarDebug failed with exit code 0.
```

```
[✓] Flutter (Channel master, 1.24.0-8.0.pre.373, on Mac OS X 10.15.7 19H15 darwin-x64, locale en)
    • Flutter version 1.24.0-8.0.pre.373 at /Users/xster/development/flutter
    • Framework revision 8de72126e9 (6 days ago), 2020-11-26 07:48:03 +0800
    • Engine revision 20caf54969
    • Dart version 2.12.0 (build 2.12.0-76.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/xster/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/xster/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.2, Build version 12B45b
    • CocoaPods version 1.10.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio
    • Android Studio at /Applications/Android Studio 4.2 Preview.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6842174)

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.51.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.16.0

[✓] Connected device (5 available)
    • Pixel 3 (mobile) • 8AXX11BS6                                • android-arm64  • Android 11 (API 30)
    • iPhone (mobile)  • cdcd493527f1f25a0ac91106540cf58186f2b4f0 • ios            • iOS 14.2
    • macOS (desktop)  • macos                                    • darwin-x64     • Mac OS X 10.15.7 19H15 darwin-x64
    • Web Server (web) • web-server                               • web-javascript • Flutter Tools
    • Chrome (web)     • chrome                                   • web-javascript • Google Chrome 87.0.4280.67

• No issues found!
```

--- a/packages/flutter_tools/lib/src/android/gradle.dart
+++ b/packages/flutter_tools/lib/src/android/gradle.dart
@@ -645,8 +645,8 @@ Future<void> buildGradleAar({
     globals.printStatus(result.stdout, wrap: false);
     globals.printError(result.stderr, wrap: false);
     throwToolExit(
-      'Gradle task $aarTask failed with exit code $exitCode.',
-      exitCode: exitCode,
+      'Gradle task $aarTask failed with exit code ${result.exitCode}.',
+      exitCode: result.exitCode,
     );
   }
   final Directory repoDirectory = getRepoDirectory(outputDirectory);

--- a/packages/flutter_tools/test/general.shard/android/gradle_test.dart
+++ b/packages/flutter_tools/test/general.shard/android/gradle_test.dart
@@ -1741,10 +1741,6 @@ plugin1=${plugin1.path}
         )
       , throwsToolExit(exitCode: 108, message: 'Gradle task assembleAarRelease failed with exit code 108.'));
 
-      expect(
-        testLogger.statusText.contains('Consuming the Module'),
-        isFalse,
-      );
     }, overrides: <Type, Generator>{
       AndroidSdk: () => mockAndroidSdk,
       AndroidStudio: () => mockAndroidStudio,

--- a/packages/flutter_tools/test/general.shard/android/gradle_test.dart
+++ b/packages/flutter_tools/test/general.shard/android/gradle_test.dart
@@ -1704,6 +1704,56 @@ plugin1=${plugin1.path}
       ProcessManager: () => mockProcessManager,
     });
 
+    testUsingContext("gradle exit code is properly passed on", () async {
+      final File manifestFile = fileSystem.file('pubspec.yaml');
+      manifestFile.createSync(recursive: true);
+      manifestFile.writeAsStringSync('''
+        flutter:
+          module:
+            androidPackage: com.example.test
+        '''
+      );
+
+      fileSystem.file('.android/gradlew').createSync(recursive: true);
+
+      fileSystem.file('.android/gradle.properties')
+          .writeAsStringSync('irrelevant');
+
+      fileSystem.file('.android/build.gradle')
+          .createSync(recursive: true);
+
+      // Let any process start. Assert after.
+      when(mockProcessManager.run(
+        any,
+        environment: anyNamed('environment'),
+        workingDirectory: anyNamed('workingDirectory'),
+      )).thenAnswer((_) async => ProcessResult(1, 108, '', ''));
+
+      fileSystem.directory('build/outputs/repo').createSync(recursive: true);
+
+      await expectLater(() async =>
+        await buildGradleAar(
+          androidBuildInfo: const AndroidBuildInfo(BuildInfo(BuildMode.release, null, treeShakeIcons: false)),
+          project: FlutterProject.current(),
+          outputDirectory: fileSystem.directory('build/'),
+          target: '',
+          buildNumber: '1.0',
+        )
+      , throwsToolExit(exitCode: 108, message: 'Gradle task assembleAarRelease failed with exit code 108.'));
+
+      expect(
+        testLogger.statusText.contains('Consuming the Module'),
+        isFalse,
+      );
+    }, overrides: <Type, Generator>{
+      AndroidSdk: () => mockAndroidSdk,
+      AndroidStudio: () => mockAndroidStudio,
+      Cache: () => cache,
+      Platform: () => android,
+      FileSystem: () => fileSystem,
+      ProcessManager: () => mockProcessManager,
+    });
+
     testUsingContext('build apk uses selected local engine,the engine abi is arm', () async {
       when(mockArtifacts.getArtifactPath(
         Artifact.flutterFramework,

--- a/packages/flutter_tools/test/general.shard/android/gradle_test.dart
+++ b/packages/flutter_tools/test/general.shard/android/gradle_test.dart
@@ -1704,7 +1704,7 @@ plugin1=${plugin1.path}
       ProcessManager: () => mockProcessManager,
     });
 
-    testUsingContext("gradle exit code is properly passed on", () async {
+    testUsingContext('gradle exit code is properly passed on', () async {
       final File manifestFile = fileSystem.file('pubspec.yaml');
       manifestFile.createSync(recursive: true);
       manifestFile.writeAsStringSync('''

